hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java

```diff
@@ -855,12 +1080,128 @@ public void breakLease(final Path path, final TracingContext tracingContext) thr
     client.breakLease(getRelativePath(path), tracingContext);
   }
 
-  public void rename(final Path source, final Path destination, TracingContext tracingContext) throws
-          AzureBlobFileSystemException {
+  public void rename(final Path source, final Path destination,
+      final AzureBlobFileSystem azureBlobFileSystem, TracingContext tracingContext) throws
+          IOException {
     final Instant startAggregate = abfsPerfTracker.getLatencyInstant();
     long countAggregate = 0;
     boolean shouldContinue;
 
+    if (getAbfsConfiguration().getPrefixMode() == PrefixMode.BLOB) {
+      LOG.debug("Rename for src: {} dst: {} for non-HNS blob-endpoint",
+          source, destination);
+      final Boolean isSrcExist;
+      final Boolean isSrcDir;
+      /*
+      * Fetch the list of blobs in the given sourcePath.
+      */
+      List<BlobProperty> srcBlobProperties = getListBlobs(source,
+          tracingContext, null, null, true);
+      final BlobProperty blobPropOnSrc;
+      if (srcBlobProperties.size() > 0) {
+        LOG.debug("src {} exists and is a directory", source);
+        isSrcExist = true;
+        isSrcDir = true;
+        /*
+        * Fetch if there is a marker-blob for the source blob.
+        */
+        BlobProperty blobPropOnSrcNullable = getBlobPropertyWithNotFoundHandling(source, tracingContext);
+        if(blobPropOnSrcNullable == null) {
+          /*
+          * There is no marker-blob, the client has to create marker blob before
+          * starting the rename.
+          */
+          //create marker file; add in srcBlobProperties;
+          LOG.debug("Source {} is a directory but there is no marker-blob", source);
+          azureBlobFileSystem.create(source);
+          Hashtable<String, String> props = new Hashtable<>();
+          props.put(HDI_ISFOLDER, "true");
+          setPathProperties(source, props, tracingContext);
+          blobPropOnSrc = new BlobProperty();
+          blobPropOnSrc.setIsDirectory(true);
+          blobPropOnSrc.setPath(source);
+        } else {
+          LOG.debug("Source {} is a directory but there is a marker-blob", source);
+          blobPropOnSrc = blobPropOnSrcNullable;
+        }
+      } else {
+        LOG.debug("source {} doesn't have any blob in its hierarchy. Checking"
+            + "if there is marker blob for it.", source);
+        blobPropOnSrc = getBlobPropertyWithNotFoundHandling(source, tracingContext);
+        if(blobPropOnSrc != null) {
+          isSrcExist = true;
+          if(blobPropOnSrc.getIsDirectory()) {
+            LOG.debug("source {} is a marker blob", source);
+            isSrcDir = true;
+          } else {
+            LOG.debug("source {} exists but is not a marker blob", source);
+            isSrcDir = false;
+          }
+        } else {
+          LOG.debug("source {} doesn't exist", source);
+          isSrcExist = false;
+          isSrcDir = false;
+        }
+      }
+      srcBlobProperties.add(blobPropOnSrc);
```

Comment:
For a folder with a large number of entries like 1Mn or so, rename is going to be held up until enumeration is complete and also the large amount of properties need to be saved in RAM. Please create a work item for post drop1 to optimize this code in a producer-consumer sort of way so that enumerate is done while parallel rename can proceed and can pause when the queue to consumer is full.

Reply:
Have added item.